### PR TITLE
GCP-TDX

### PR DESCRIPTION
**Pull request by phochard**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -182,19 +182,30 @@ jobs:
         with:
           token: ${{ secrets.GITHUB_TOKEN }}
 
-  build-test-gcp-image:
+  build-test-gcp-sev-image:
     strategy:
       matrix:
         distrib: [ubuntu, rhel]
     name: GCP image (${{ matrix.distrib }})
```
Comment (Manuthor) [inline .github/workflows/ci.yml:189]: If you don't mind (clearer on github pipeline):

```suggestion
    name: ${{ matrix.distrib }} -> GCP image
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -182,19 +182,30 @@ jobs:
...
     secrets: inherit
-    uses: ./.github/workflows/gcp_image.yml
+    uses: ./.github/workflows/gcp_sev_image.yml
+    with:
+      distrib: ${{ matrix.distrib }}
+
+  build-test-gcp-tdx-image:
+    strategy:
+      matrix:
+        distrib: [ubuntu]
```
Comment (Manuthor) [inline .github/workflows/ci.yml:199]: for my knowledge, should we also test on redhat?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -182,19 +182,30 @@ jobs:
...
+    name: GCP image (${{ matrix.distrib }})
```
Comment (Manuthor) [inline .github/workflows/ci.yml:200]:
```suggestion
    name: ${{ matrix.distrib }} -> GCP image
```

File: .github/workflows/gcp_tdx_image.yml [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        required: true
+        type: string
+
+jobs:
+  build-image:
+    name: Packer build
+    runs-on: ubuntu-22.04
+    permissions:
+      contents: "read"
+      id-token: "write"
```
Comment (Manuthor) [inline .github/workflows/gcp_tdx_image.yml:16]:
```suggestion
      contents: read
      id-token: write
```

**Review by Manuthor — COMMENTED** [review]
just small remarks. Also, do you think we could merge `gcp_tdx_image.yml` and `gcp_sev_image.yml` in 1 reusable workflow (with some parameters).
Great job, thank you!

File: .github/workflows/ci.yml (hunk 2 above)
Comment (phochard) [reply .github/workflows/ci.yml:199]: currently there is no redhat tdx compatible source image available on GCP 
```
remyfrancois:~$ gcloud alpha compute images list \
   --project=tdx-guest-images \
   --filter="guestOsFeatures[].type:(TDX_CAPABLE)"
NAME                       PROJECT           FAMILY             DEPRECATED  STATUS
cos-109-tdx-base-20231016  tdx-guest-images  cos-tdx                        READY
ubuntu-2204-tdx-v20231011  tdx-guest-images  ubuntu-2204-lts                READY
ubuntu-2304-tdx-v20231011  tdx-guest-images  ubuntu-2304-amd64              READY
```

that's why we choose to duplicate the job, it failed with the matrix including redhat

Comment (Manuthor) [reply .github/workflows/ci.yml:199]: thank you for the explanation

File: .github/workflows/gcp_tdx_image.yml [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+          ./target/release/cosmian_vm --url https://${IP_ADDR}:5355 --allow-insecure-tls snapshot
+          ./target/release/cosmian_vm --url https://${IP_ADDR}:5355 --allow-insecure-tls verify --snapshot ./cosmian_vm.snapshot
+
+      - name: Stop and delete GCP instance
+        if: success() || failure()
+        env:
+          CI_INSTANCE: ${{ needs.build-image.outputs.ci_instance }}
+        run: |
+          set +e
+          OUTPUT=$(gcloud alpha compute instances describe $CI_INSTANCE --format='get(status)' --zone=us-central1-a)
```
[the remark this answers is not in the document]
Comment (phochard) [reply .github/workflows/gcp_tdx_image.yml:133]: I agree with you.
This part have been written by @grydz , we should ask him if this command is really necessary.

File: packer/gcp-tdx-ubuntu.pkr.hcl [hunk 6]
```diff
@@ -7,12 +7,12 @@ gcloud alpha compute --project=intel-enclaves images create ubuntu-2204-tdx-v202
...
 
 locals {
-  ubuntu_ami_name = "${var.prefix}-cosmian-vm-tdx-ubuntu-{{timestamp}}"
+  ubuntu_ami_name = "${var.prefix}-cosmian-vm-tdx-ubuntu"
 }
 
 variable "project_id" {
   type    = string
-  default = "intel-enclaves"
+  default = "cosmian-dev"
```
[the remark this answers is not in the document]
Comment (phochard) [reply packer/gcp-tdx-ubuntu.pkr.hcl:15]: Again, I agree with you.
We didn't get why @grydz keep using other project than `cosmian-dev`.
That is why we used `cosmian-dev` in the tdx part.

File: .github/workflows/gcp_image.yml [hunk 7]
```diff
@@ -95,18 +129,38 @@ jobs:
...
+                                                                              --image="$IMAGE_NAME" \
+                                                                              --image-project=${{ inputs.project }} \
+                                                                              --project ${{ inputs.project }} \
+                                                                              --tags ssh-full,backend-rust-full,http-server,https-server,cosmian-vm-agent \
+                                                                              --metadata-from-file=startup-script=resources/scripts/gcp-start-script.sh \
+                                                                              --max-run-duration=10m \
+                                                                              --instance-termination-action=DELETE
+              IP_ADDR=$(gcloud ${{ inputs.mode }} compute instances describe $CI_INSTANCE --format='get(networkInterfaces[0].accessConfigs[0].natIP)' --zone=${{ inputs.zone }})
+              echo "IP_ADDR=${IP_ADDR}" >> "$GITHUB_OUTPUT"
+          fi
```
Comment (Manuthor) [inline .github/workflows/gcp_image.yml:163]: any chance to merge the 2 blocks in 1?
(remove the if-else statements)

File: .github/workflows/gcp_image.yml [hunk 8]
```diff
@@ -148,13 +202,11 @@ jobs:
...
-          EXIT_CODE="$?"
-          if [ "$EXIT_CODE" -eq 0 ]; then
-            gcloud beta compute instances delete $CI_INSTANCE --zone europe-west4-a \
-                                                              --project $GCP_DEV_PROJECT
-          fi
+            set +e
+            OUTPUT=$(gcloud ${{ inputs.mode }} compute instances describe $CI_INSTANCE --format='get(status)' --zone=${{ inputs.zone }})
+            EXIT_CODE="$?"
+            gcloud ${{ inputs.mode }} compute instances delete $CI_INSTANCE --zone ${{ inputs.zone }} --project ${{ inputs.project }}
+
```
Comment (Manuthor) [inline .github/workflows/gcp_image.yml:209]: We have done this for the KMS, what do you think about doing the equivalent for cosmian VM (in order to delete tmp images)
```
      - name: Delete GCP CI image
        if: startsWith(github.ref, 'refs/tags') != true
        env:
          IMAGE_NAME: ${{ needs.build-kms-cosmian-vm-image.outputs.image_name }}
        run: |
          set -ex
          gcloud beta compute images delete --quiet $IMAGE_NAME
```

File: .github/workflows/gcp_image.yml [hunk 9]
```diff
@@ -31,7 +52,11 @@ jobs:
         run: |
           TIMESTAMP="$(date -u +'%Y%m%d%H%M%S')"
           echo "TIMESTAMP=$TIMESTAMP" >> "$GITHUB_OUTPUT"
+          if [ "${{ inputs.techno }}" == "sev" ]; then
           echo "IMAGE_NAME=alpha-$TIMESTAMP-cosmian-vm-sev-${{ inputs.distrib }}" >> "$GITHUB_OUTPUT"
+          elif [ "${{ inputs.techno }}" == "tdx" ]; then
+          echo "IMAGE_NAME=alpha-$TIMESTAMP-cosmian-vm-tdx-${{ inputs.distrib }}" >> "$GITHUB_OUTPUT"
+          fi
```
Comment (Manuthor) [inline .github/workflows/gcp_image.yml:59]: could we merge this in 1 block? (by using ${{ inputs.techno }})?
 ```suggestion
          echo "IMAGE_NAME=alpha-$TIMESTAMP-cosmian-vm-${{ inputs.techno }}-${{ inputs.distrib }}" >> "$GITHUB_OUTPUT"
```

File: .github/workflows/gcp_image.yml [hunk 10]
```diff
@@ -53,7 +78,12 @@ jobs:
           packer plugins install github.com/hashicorp/ansible
 
       - name: Init packer
-        run: cd packer/ && packer init gcp-sev-${{ inputs.distrib }}.pkr.hcl
+        run: |
+          if [ "${{ inputs.techno }}" == "sev" ]; then
+            cd packer/ && packer init gcp-sev-${{ inputs.distrib }}.pkr.hcl
+          elif [ "${{ inputs.techno }}" == "tdx" ]; then
+            cd packer/ && packer init gcp-tdx-${{ inputs.distrib }}.pkr.hcl
+          fi
```
Comment (Manuthor) [inline .github/workflows/gcp_image.yml:86]: same here

File: .github/workflows/gcp_image.yml [hunk 11]
```diff
@@ -65,15 +95,19 @@ jobs:
         env:
           TIMESTAMP: ${{ steps.env.outputs.TIMESTAMP }}
         run: |
-          cd packer/ && packer build -var "prefix=alpha-$TIMESTAMP" gcp-sev-${{ inputs.distrib }}.pkr.hcl
+          if [ "${{ inputs.techno }}" == "sev" ]; then
+            cd packer/ && packer build -var "prefix=alpha-$TIMESTAMP" gcp-sev-${{ inputs.distrib }}.pkr.hcl
+          elif [ "${{ inputs.techno }}" == "tdx" ]; then
+            cd packer/ && packer build -var "prefix=alpha-$TIMESTAMP" gcp-tdx-${{ inputs.distrib }}.pkr.hcl
+          fi
```
Comment (Manuthor) [inline .github/workflows/gcp_image.yml:102]: same here

File: packer/gcp-tdx-ubuntu.pkr.hcl [hunk 12]
```diff
@@ -111,12 +119,12 @@ build {
   }
 
   provisioner "file" {
-    source      = "./target/release/cosmian_vm_agent"
+    source      = "./../target/release/cosmian_vm_agent"
```
Comment (ThibsG) [inline packer/gcp-tdx-ubuntu.pkr.hcl:122]:
```suggestion
    source      = "../target/release/cosmian_vm_agent"
```

File: packer/gcp-tdx-ubuntu.pkr.hcl [hunk 13]
```diff
@@ -111,12 +119,12 @@ build {
...
     destination = "/tmp/"
   }
 
   provisioner "file" {
-    source      = "./target/release/cosmian_certtool"
+    source      = "./../target/release/cosmian_certtool"
```
Comment (ThibsG) [inline packer/gcp-tdx-ubuntu.pkr.hcl:127]:
```suggestion
    source      = "../target/release/cosmian_certtool"
```

File: .github/workflows/gcp_image.yml [hunk 14]
```diff
@@ -1,11 +1,35 @@
-name: Build and test image for GCP
+name: Build and test image for GCP SEV
```
Comment (ThibsG) [inline .github/workflows/gcp_image.yml:1]:
```suggestion
name: Build and test image for GCP SEV and TDX
```
it is not only SEV right?

**Review by ThibsG — APPROVED** [review]

**Review by grydz — APPROVED** [review]

**Pull request merged**